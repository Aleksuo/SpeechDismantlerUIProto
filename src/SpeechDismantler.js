import React, { Component } from 'react'
import {Hidden } from '@material-ui/core'
import MiniDrawer from "./common/MiniDrawer"
import MobileDrawer from "./common/MobileDrawer"

//import views
import HomePage from "./views/homepage/HomePage"

// import { VictoryBar, VictoryTheme, VictoryChart, VictoryPie } from 'victory';
import openSocket from 'socket.io-client'
import PropTypes from 'prop-types'

import { downsampleBuffer } from './utils/AudioUtils.js'
import { estimateStartTime } from './utils/GeneralUtils'


let AudioContext
let context
let processor
let input
let globalStream

const initialState = {
	isRecording: false,
	elapsed: 0,
	transcript: [],
	interim: "",
	left: false,
}

class SpeechDismantler extends Component {
	constructor(props) {
		super(props)
		let { server_address } = props
		this.bufferSize = 2048
		this.socket = openSocket(server_address)
		this.state = initialState

		this.socket.on('connect', () => {
			this.socket.emit('join', 'Server Connected to Client')
		})

		this.socket.on('speechData', (data) => {
			const final = undefined || data.results[0].isFinal
			const result = data.results[0].alternatives[0].words
			const transcript = data.results[0].alternatives[0].transcript
			if (final === false) {
				this.setState({
					interim: transcript,
				})
			} else {
				var newTranscript = this.state.transcript.slice(0)
				var sentence = {
					startTime: 0,
					endTime: this.state.elapsed,
					words: result
				}
				const startTime = estimateStartTime(sentence)
				//console.log(startTime)
				sentence.startTime = startTime
				newTranscript.push(sentence)
				this.setState({
					transcript: newTranscript,
				})
			}
		})

		window.onbeforeunload = () => {
			if (this.state.isRecording) { this.socket.emit('endGoogleCloudStream', '') }
		}
	}

	tick = () => {
		const newElapsed = this.state.elapsed + (new Date() - this.last)
		this.setState({ elapsed: newElapsed })
		this.last = new Date()
	}

	reset = () => {
		if (this.state.isRecording) {
			this.stopRecording()
		}
		this.setState(initialState, clearInterval(this.timer))
	}


	toggleRecord = () => {
		const newIsRecording = !this.state.isRecording // state might not be updated if it is read after setState
		this.setState({
			isRecording: newIsRecording,
		},
			newIsRecording
				? () => {
					this.last = new Date()
					this.timer = setInterval(this.tick, 100)
					return this.handleListen()
				}
				: () => {
					clearInterval(this.timer)
					return this.stopRecording()
				})

	}

	streamAudioData = (e) => {
		const left = e.inputBuffer.getChannelData(0)
		const left16 = downsampleBuffer(left, 44100, 16000)
		this.socket.emit('binaryData', left16)
	}

	handleListen = () => {
		this.socket.emit('startGoogleCloudStream', '') // init socket Google Speech Connection
		AudioContext = window.AudioContext || window.webkitAudioContext
		context = new AudioContext()
		processor = context.createScriptProcessor(this.bufferSize, 1, 1)
		processor.connect(context.destination)
		context.resume()

		const handleSuccess = (stream) => {
			globalStream = stream
			input = context.createMediaStreamSource(stream)
			input.connect(processor)
			processor.onaudioprocess = (e) => {
				this.streamAudioData(e)
			}
		}

		navigator.mediaDevices.getUserMedia({ audio: true, video: false })
			.then(handleSuccess)
	}

	stopRecording = () => {
		this.socket.emit('endGoogleCloudStream', '')


		const track = globalStream.getTracks()[0]
		track.stop()

		input.disconnect(processor)
		processor.disconnect(context.destination)
		context.close().then(() => {
			input = null
			processor = null
			context = null
			AudioContext = null
		})
	}

	//UI CODE STARTS HERE*/
	render() {
		return (
			<div>
				<div>
					<Hidden smDown>
						<MiniDrawer />
					</Hidden>
					<Hidden mdUp>
						<MobileDrawer />
					</Hidden>
				</div>
				<div>
					<HomePage state={this.state} toggleRecord={this.toggleRecord} reset={this.reset} />
				</div>
			</div >
		)
	}
}

SpeechDismantler.propTypes = {
	server_address: PropTypes.string
}

export default SpeechDismantler
