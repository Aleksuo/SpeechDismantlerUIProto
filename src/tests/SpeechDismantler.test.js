/* eslint-disable */
import React from "react"
import Enzyme, {shallow, mount,render} from "enzyme"
import SpeechDismantler from "../SpeechDismantler"
import Adapter from "enzyme-adapter-react-16"

Enzyme.configure({adapter: new Adapter()})


describe("SpeechDismantler component", () =>{
    test("renders", () =>{ 
        const wrapper = shallow(<SpeechDismantler server_address={"..."}/>) 
        expect(wrapper).toMatchSnapshot()
    })
})