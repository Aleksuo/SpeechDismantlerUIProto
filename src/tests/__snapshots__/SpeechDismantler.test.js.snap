// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeechDismantler component renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpeechDismantler
    server_address="..."
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <Hidden
            implementation="js"
            lgDown={false}
            lgUp={false}
            mdDown={false}
            mdUp={false}
            smDown={true}
            smUp={false}
            xlDown={false}
            xlUp={false}
            xsDown={false}
            xsUp={false}
          >
            <WithStyles(MiniDrawer)
              setView={[Function]}
            />
          </Hidden>
          <Hidden
            implementation="js"
            lgDown={false}
            lgUp={false}
            mdDown={false}
            mdUp={true}
            smDown={false}
            smUp={false}
            xlDown={false}
            xlUp={false}
            xsDown={false}
            xsUp={false}
          >
            <MobileDrawer
              setView={[Function]}
            />
          </Hidden>
        </div>,
        <div>
          <HomePage
            reset={[Function]}
            state={
              Object {
                "elapsed": 0,
                "interim": "",
                "isRecording": false,
                "left": false,
                "transcript": Array [],
                "view": 0,
                "volumes": Array [],
              }
            }
            toggleRecord={[Function]}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={true}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <WithStyles(MiniDrawer)
                setView={[Function]}
              />
            </Hidden>,
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={true}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <MobileDrawer
                setView={[Function]}
              />
            </Hidden>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <WithStyles(MiniDrawer)
                setView={[Function]}
              />,
              "implementation": "js",
              "lgDown": false,
              "lgUp": false,
              "mdDown": false,
              "mdUp": false,
              "smDown": true,
              "smUp": false,
              "xlDown": false,
              "xlUp": false,
              "xsDown": false,
              "xsUp": false,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "setView": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MobileDrawer
                setView={[Function]}
              />,
              "implementation": "js",
              "lgDown": false,
              "lgUp": false,
              "mdDown": false,
              "mdUp": true,
              "smDown": false,
              "smUp": false,
              "xlDown": false,
              "xlUp": false,
              "xsDown": false,
              "xsUp": false,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "setView": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <HomePage
            reset={[Function]}
            state={
              Object {
                "elapsed": 0,
                "interim": "",
                "isRecording": false,
                "left": false,
                "transcript": Array [],
                "view": 0,
                "volumes": Array [],
              }
            }
            toggleRecord={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "reset": [Function],
            "state": Object {
              "elapsed": 0,
              "interim": "",
              "isRecording": false,
              "left": false,
              "transcript": Array [],
              "view": 0,
              "volumes": Array [],
            },
            "toggleRecord": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={false}
              smDown={true}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <WithStyles(MiniDrawer)
                setView={[Function]}
              />
            </Hidden>
            <Hidden
              implementation="js"
              lgDown={false}
              lgUp={false}
              mdDown={false}
              mdUp={true}
              smDown={false}
              smUp={false}
              xlDown={false}
              xlUp={false}
              xsDown={false}
              xsUp={false}
            >
              <MobileDrawer
                setView={[Function]}
              />
            </Hidden>
          </div>,
          <div>
            <HomePage
              reset={[Function]}
              state={
                Object {
                  "elapsed": 0,
                  "interim": "",
                  "isRecording": false,
                  "left": false,
                  "transcript": Array [],
                  "view": 0,
                  "volumes": Array [],
                }
              }
              toggleRecord={[Function]}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Hidden
                implementation="js"
                lgDown={false}
                lgUp={false}
                mdDown={false}
                mdUp={false}
                smDown={true}
                smUp={false}
                xlDown={false}
                xlUp={false}
                xsDown={false}
                xsUp={false}
              >
                <WithStyles(MiniDrawer)
                  setView={[Function]}
                />
              </Hidden>,
              <Hidden
                implementation="js"
                lgDown={false}
                lgUp={false}
                mdDown={false}
                mdUp={true}
                smDown={false}
                smUp={false}
                xlDown={false}
                xlUp={false}
                xsDown={false}
                xsUp={false}
              >
                <MobileDrawer
                  setView={[Function]}
                />
              </Hidden>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <WithStyles(MiniDrawer)
                  setView={[Function]}
                />,
                "implementation": "js",
                "lgDown": false,
                "lgUp": false,
                "mdDown": false,
                "mdUp": false,
                "smDown": true,
                "smUp": false,
                "xlDown": false,
                "xlUp": false,
                "xsDown": false,
                "xsUp": false,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "setView": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <MobileDrawer
                  setView={[Function]}
                />,
                "implementation": "js",
                "lgDown": false,
                "lgUp": false,
                "mdDown": false,
                "mdUp": true,
                "smDown": false,
                "smUp": false,
                "xlDown": false,
                "xlUp": false,
                "xsDown": false,
                "xsUp": false,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "setView": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <HomePage
              reset={[Function]}
              state={
                Object {
                  "elapsed": 0,
                  "interim": "",
                  "isRecording": false,
                  "left": false,
                  "transcript": Array [],
                  "view": 0,
                  "volumes": Array [],
                }
              }
              toggleRecord={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "reset": [Function],
              "state": Object {
                "elapsed": 0,
                "interim": "",
                "isRecording": false,
                "left": false,
                "transcript": Array [],
                "view": 0,
                "volumes": Array [],
              },
              "toggleRecord": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
